crates/miden-objects/src/block/nullifier_tree.rs

```diff
@@ -334,4 +531,157 @@ mod tests {
         assert_matches!(err, NullifierTreeError::NullifierAlreadySpent(nullifier) if nullifier == nullifier1);
     }
+
+    #[cfg(feature = "std")]
+    #[test]
+    fn large_smt_backend_basic_operations() {
+        use miden_crypto::merkle::{LargeSmt, MemoryStorage};
+
+        // Create test data
+        let nullifier1 = Nullifier::dummy(1);
+        let nullifier2 = Nullifier::dummy(2);
+        let nullifier3 = Nullifier::dummy(3);
+
+        let block1 = BlockNumber::from(1);
+        let block2 = BlockNumber::from(2);
+        let block3 = BlockNumber::from(3);
+
+        // Create NullifierTree with LargeSmt backend
+        let tree = LargeSmt::with_entries(
+            MemoryStorage::default(),
+            [
+                (nullifier1.as_word(), super::block_num_to_nullifier_leaf_value(block1)),
+                (nullifier2.as_word(), super::block_num_to_nullifier_leaf_value(block2)),
+            ],
+        )
+        .map(NullifierTree::new_unchecked)
+        .unwrap();
+
+        // Test basic operations
+        assert_eq!(tree.num_nullifiers(), 2);
+        assert_eq!(tree.get_block_num(&nullifier1).unwrap(), block1);
+        assert_eq!(tree.get_block_num(&nullifier2).unwrap(), block2);
+
+        // Test opening
+        let _witness1 = tree.open(&nullifier1);
+
+        // Test mutations
+        let mut tree_mut = LargeSmt::with_entries(
+            MemoryStorage::default(),
+            [
+                (nullifier1.as_word(), super::block_num_to_nullifier_leaf_value(block1)),
+                (nullifier2.as_word(), super::block_num_to_nullifier_leaf_value(block2)),
+            ],
+        )
+        .map(NullifierTree::new_unchecked)
+        .unwrap();
+        tree_mut.mark_spent(nullifier3, block3).unwrap();
+        assert_eq!(tree_mut.num_nullifiers(), 3);
+        assert_eq!(tree_mut.get_block_num(&nullifier3).unwrap(), block3);
+
+        // Verify original tree unchanged
+        assert_eq!(tree.num_nullifiers(), 2);
+    }
+
+    #[cfg(feature = "std")]
+    #[test]
+    fn large_smt_backend_nullifier_already_spent() {
+        use miden_crypto::merkle::{LargeSmt, MemoryStorage};
+
+        let nullifier1 = Nullifier::dummy(1);
+
+        let block1 = BlockNumber::from(1);
+        let block2 = BlockNumber::from(2);
+
+        let mut tree = NullifierTree::new_unchecked(
+            LargeSmt::with_entries(
+                MemoryStorage::default(),
+                [(nullifier1.as_word(), super::block_num_to_nullifier_leaf_value(block1))],
+            )
+            .unwrap(),
+        );
+
+        assert_eq!(tree.get_block_num(&nullifier1).unwrap(), block1);
+
+        // Create a new tree for the first test
+        let mut tree_test1 = NullifierTree::new_unchecked(
+            LargeSmt::with_entries(
+                MemoryStorage::default(),
+                [(nullifier1.as_word(), super::block_num_to_nullifier_leaf_value(block1))],
+            )
+            .unwrap(),
+        );
```

Comment:
Similar to the other test, I think I don't understand why we create a second tree instead of testing against the first one?

Reply:
Refactored